src/builtins/fish_indent.rs

```diff
@@ -0,0 +1,1149 @@
+                self.visit_semi_nl(token);
+                return;
+            }
+            self.emit_node_text(node);
+            return;
+        }
+        match node.typ() {
+            Type::argument | Type::variable_assignment => {
+                self.emit_node_text(node);
+            }
+            Type::redirection => {
+                self.visit_redirection(node.as_redirection().unwrap());
+            }
+            Type::maybe_newlines => {
+                self.visit_maybe_newlines(node.as_maybe_newlines().unwrap());
+            }
+            Type::begin_header => {
+                // 'begin' does not require a newline after it, but we insert one.
+                node.accept(self, false);
+                self.visit_begin_header();
+            }
+            _ => {
+                // For branch and list nodes, default is to visit their children.
+                if [Category::branch, Category::list].contains(&node.category()) {
+                    node.accept(self, false);
+                    return;
+                }
+                panic!("unexpected node type");
+            }
+        }
+    }
+}
+
+/// Return whether a character at a given index is escaped.
+/// A character is escaped if it has an odd number of backslashes.
+fn char_is_escaped(text: &wstr, idx: usize) -> bool {
+    count_preceding_backslashes(text, idx) % 2 == 1
+}
+
+pub fn main() {
+    PROGRAM_NAME.set(L!("fish_indent")).unwrap();
+    panic_handler(throwing_main)
+}
+
+fn throwing_main() -> i32 {
+    // TODO: Duplicated with fish_key_reader
+    use crate::io::FdOutputStream;
+    use crate::io::IoChain;
+    use crate::io::OutputStream::Fd;
+    use libc::{STDERR_FILENO, STDOUT_FILENO};
+
+    topic_monitor_init();
+    threads::init();
+
+    let mut out = Fd(FdOutputStream::new(STDOUT_FILENO));
+    let mut err = Fd(FdOutputStream::new(STDERR_FILENO));
+    let io_chain = IoChain::new();
+    let mut streams = IoStreams::new(&mut out, &mut err, &io_chain);
+    // Using the user's default locale could be a problem if it doesn't use UTF-8 encoding. That's
+    // because the fish project assumes Unicode UTF-8 encoding in all of its scripts.
+    //
+    {
+        let s = CString::new("").unwrap();
+        unsafe { libc::setlocale(LC_ALL, s.as_ptr()) };
+    }
+    env_init(None, true, false);
+    let args: Vec<WString> = std::env::args_os()
+        .map(|osstr| str2wcstring(osstr.as_bytes()))
+        .collect();
+
+    // Only set these here so you can't set them via the builtin.
+    if let Some(features_var) = EnvStack::globals().get(L!("fish_features")) {
+        for s in features_var.as_list() {
+            future_feature_flags::set_from_string(s.as_utfstr());
+        }
+    }
+
+    do_indent(&mut streams, args)
+}
+
+pub fn fish_indent(_parser: &Parser, streams: &mut IoStreams, args: &mut [&wstr]) -> Option<c_int> {
+    let args = args.iter_mut().map(|x| x.to_owned()).collect();
+    Some(do_indent(streams, args))
```

Comment:
`future_feature_flags::set_from_string(s.as_utfstr());` is not necessary when running as builtin, fish already does that

Reply:
Yes, it's in throwing_main, which is only used when run as a command.